### DISCUSSION TITLE
Mapped dataloader, how to expose the correct value?

### DISCUSSION THREAD

**Discussion by jarlehansen** [ExpediaGroup/graphql-kotlin#1029] [Q&A] [answered]
We are using data loaders extensively, and for the regular data loaders it works great with graphql-kotlin. We return a `CompletableFuture<V>` and it gets dispatched.

However, we need to use the mapped data loader (`DataLoader.newMappedDataLoader`) to map the key to the value(s). The return value for this type of data loader is `Map<K, V>`, which means it will be `CompletableFuture<Map<K, V>>`. When I add this to a graphql-kotlin query class, it fails because it says that `Map` is not a graphql type.

Is there any examples of doing this?
Does it not support this type of data loader, or am I doing something wrong?

**Comment by jarlehansen** [comment, marked as answer]
I think I might have done something wrong in the load function, but there is something I do not understand. If we have a simple return value, such as `CompletableFuture<MyValue>` it works. However, if I need a list of values (`CompletableFuture<List<MyValue>>` it gives this error: 

```
type mismatch error, expected type LIST got class java.util.concurrent.CompletableFuture
```

**Reply by smyrick** [reply]
Hmm interesting on the `CompletableFuture<List<MyValue>>` as that should be a valid return type. I can check if that throws an error with a basic test.

On the `CompletableFuture<Map<K, V>>`, that is indeed an invalid return type so can you explain a little bit more how you are using the  `MappedDataLoader` as it still should return a single Dataloader

https://javadoc.io/doc/com.graphql-java/java-dataloader/2.2.1/org/dataloader/DataLoader.html

**Reply by smyrick** [reply]
The `CompletableFuture<List<String>>` is valid.

```kotlin
@Test
fun validateSchema() {
    val queries = listOf(TopLevelObject(SimpleQuery()))
    val schema = toSchema(testSchemaConfig, queries)
    assertNotNull(schema)
    assertEquals("String!", schema.queryType.getFieldDefinition("getString").type.toString())
    assertEquals("String!", schema.queryType.getFieldDefinition("getDataFetcherResultString").type.toString())
    assertEquals("[String!]!", schema.queryType.getFieldDefinition("getListString").type.toString())
}

class SimpleQuery {
    private val string = "hello"

    fun getString(): CompletableFuture<String> = CompletableFuture.completedFuture(string)
    fun getDataFetcherResultString(): CompletableFuture<DataFetcherResult<String>> =
        CompletableFuture.completedFuture(DataFetcherResult.newResult<String>().data(string).build())
    fun getListString(): CompletableFuture<List<String>> = CompletableFuture.completedFuture(listOf(string))
}
```

**Reply by smyrick** [reply]
https://github.com/ExpediaGroup/graphql-kotlin/pull/1032

**Reply by jarlehansen** [reply]
Thanks for the examples, really useful.
What about `CompletableFuture<List<DataFetcherResult<String>>>`? I get an error on this:
```
com.expediagroup.graphql.exceptions.TypeNotSupportedException: Cannot convert graphql.execution.DataFetcherResult
```

Shouldnt this be allowed if you have more than one value pr key?

**Reply by smyrick** [reply]
It would be `CompletableFuture<DataFetcherResult<List<String>>>`

 `CompletableFuture` and `DataFetcherResult` are just wrapping types, whereas `List<String>` is the actual type exposed in the schema. The errors field in the single `DataFetcherResult` is a list of errors you can append to
